### PR TITLE
Only add default services once in WebApplicationBuilder

### LINKED ISSUE #33934: WebApplicationBuilder adds default services twice
After my [change](https://github.com/dotnet/aspnetcore/pull/33800) got merged Swashbuckle broke when using the `WebApplicationBuilder` and it's because we call  ConfigureWebHostDefaults twice: 
1. https://github.com/dotnet/aspnetcore/blob/348b810d286fd2258aa763d6eda667a83ff972dc/src/DefaultBuilder/src/WebApplicationBuilder.cs#L52
2. https://github.com/dotnet/aspnetcore/blob/348b810d286fd2258aa763d6eda667a83ff972dc/src/DefaultBuilder/src/WebApplicationBuilder.cs#L111

Assuming it was fine to do so but it isn't when using API explorer because we end up adding the same provider twice which makes swashbuckle (and I assume other code) unhappy. Swagger/OpenAPI is currently broken because of this.

cc @halter73 @LadyNaggaga @bradygaster

### PR DESCRIPTION
**Pull request by halter73**
Fixes #33934

### CODE REVIEW

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 1]
```diff
@@ -533,9 +532,50 @@ public async Task WebApplicationBuilder_StartupFilterCanAddTerminalMiddleware()
...
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IConfigureOptions<LoggerFactoryOptions>>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IWebHostEnvironment>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IOptionsChangeTokenSource<HostFilteringOptions>>>());
+            Assert.Single(app.Services.GetRequiredService<IEnumerable<IServer>>());
+        }
+
+        [Fact]
+        public void WebApplicationBuilder_EnablesServiceScopeValidationByDefaultInDevelopment()
+        {
+            // The environment cannot be reconfigured after the builder is created currently.
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:558]: Not part of this PR but we should talk about ways to throw if `UseEnvironment` is used after this call (and maybe we need a first class API).

File: src/DefaultBuilder/src/WebHostEnvironment.cs [hunk 2]
```diff
@@ -56,17 +56,15 @@ internal void ReadConfigurationSettings(IConfiguration configuration)
...
             genericWebHostBuilder.UseSetting(WebHostDefaults.EnvironmentKey, EnvironmentName);
             genericWebHostBuilder.UseSetting(WebHostDefaults.ContentRootKey, ContentRootPath);
             genericWebHostBuilder.UseSetting(WebHostDefaults.WebRootKey, WebRootPath);
 
-            genericWebHostBuilder.ConfigureAppConfiguration((context, builder) =>
-            {
-                CopyPropertiesTo(context.HostingEnvironment);
-            });
+            var context = (WebHostBuilderContext)hostBuilder.Properties[typeof(WebHostBuilderContext)];
+            CopyPropertiesTo(context.HostingEnvironment);
```
Comment (davidfowl) [inline src/DefaultBuilder/src/WebHostEnvironment.cs:67]: Do you even need to do this?

Comment (halter73) [reply src/DefaultBuilder/src/WebHostEnvironment.cs:67]: If not, we get test failures [here](https://github.com/dotnet/aspnetcore/blob/dd476b1fdfd851dcfe0cab21c5084f05059de6bb/src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs#L189) and other similar places since WebApplication.Environment pulls from the IServiceProvider and we no longer register our IWebHostEnvironment as a service.

**Comment by davidfowl** [comment]
I debugged the issue, the IWebHostEnvironment.ApplicationName is incorrect and it results in not being able to find controllers or views.

File: src/DefaultBuilder/src/BootstrapHostBuilder.cs [hunk 3]
```diff
@@ -103,6 +118,16 @@ internal void RunConfigurationCallbacks()
             }
 
             _environment.ApplyConfigurationSettings(_configuration);
+
```
Comment (davidfowl) [inline src/DefaultBuilder/src/BootstrapHostBuilder.cs:121]: Why do we `ApplyConfigurationSettings` twice when nothing is observing changes to `_environment` between these 2 calls.?

File: src/DefaultBuilder/src/WebApplicationBuilder.cs [hunk 4]
```diff
@@ -34,31 +37,30 @@ internal WebApplicationBuilder(Assembly? callingAssembly, string[]? args = null)
...
-
-            Logging = new LoggingBuilder(Services);
-            WebHost = _deferredWebHostBuilder = new ConfigureWebHostBuilder(Configuration, _environment, Services);
-            Host = _deferredHostBuilder = new ConfigureHostBuilder(Configuration, _environment, Services);
+            _bootstrapHostBuilder = new BootstrapHostBuilder(Configuration, _environment, Services);
+            _bootstrapHostBuilder.ConfigureDefaults(args);
+            _bootstrapHostBuilder.ConfigureWebHostDefaults(webHostBuilder =>
+            {
+                // Runs inline.
+                webHostBuilder.Configure(ConfigureApplication);
```
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:49]:
```suggestion
                webHostBuilder.Configure(ConfigureApplication);

                // We need to override the application name since the call to Configure will set it to
                // be the calling assembly's name.
                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, _environment.ApplicationName);
```

Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplicationBuilder.cs:49]: @halter73 this should fix the issue.

File: src/DefaultBuilder/src/WebApplicationBuilder.cs [hunk 5]
```diff
@@ -116,10 +125,36 @@ public WebApplication Build()
...
+                // We've only added services configured by the GenericWebHostBuilder and WebHost.ConfigureWebDefaults
+                // at this point. HostBuilder news up a new ServiceCollection in HostBuilder.Build() we haven't seen
+                // until now, so we cannot clear these services even though some are redundant because
+                // we called ConfigureWebHostDefaults on both the _deferredHostBuilder and _hostBuilder.
+
+                // Ideally, we'd only call _hostBuilder.ConfigureWebHost(ConfigureWebHost) instead of
+                // _hostBuilder.ConfigureWebHostDefaults(ConfigureWebHost) to avoid some duplicate service descriptors,
+                // but we want to add services in the WebApplicationBuilder constructor so code can inspect
+                // WebApplicationBuilder.Services. At the same time, we want to be able which services are loaded
+                // to react to config changes (e.g. ForwardedHeadersStartupFilter).
```
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:141]:
```suggestion
```

**Comment by davidfowl** [comment]
This PR gave me clarity on how we should fix the remaining issues with the WebApplicationBuilder.

**Review by davidfowl — APPROVED** [review]

File: src/DefaultBuilder/src/BootstrapHostBuilder.cs (hunk 3 above)
Comment (halter73) [reply src/DefaultBuilder/src/BootstrapHostBuilder.cs:121]: This isn't new. ConfigureHostConfiguration callbacks to read environment variables (e.g. `DOTNET_` and `ASPNETCORE_`) that need to be reacted to before running ConfigureAppConfiguration callbacks to read config that's dependent on the environment name configured by the environment variables (e.g. `appsettings.{env.EnvironmentName}.json` and user secrets).

This is the last call to `ApplyConfigurationSettings` before the WebApplicationBuilder ctor exits, so it's to ensure app code in Program.cs can observe any changes made to the environment in appsettings mostly.

Comment (davidfowl) [reply src/DefaultBuilder/src/BootstrapHostBuilder.cs:121]: After playing with this a bit I think we only need to read hosting configuration once at the beginning and blocking it after that. We should also get rid of our custom IWebHostEnvironment and should use the one configured in the `IHostBuilder.Properties[typeof(WebHostContext)]`

Comment (halter73) [reply src/DefaultBuilder/src/BootstrapHostBuilder.cs:121]: Agreed!
